session/schema_amender.go

```diff
@@ -331,38 +323,77 @@ func (a *amendOperationAddIndexInfo) String() string {
 	return res
 }
 
-func (a *amendOperationDeleteOldIndex) genMutations(ctx context.Context, sctx sessionctx.Context,
-	commitMutations tikv.CommitterMutations, kvMap *rowKvMap, resAddMutations *tikv.CommitterMutations) error {
+func (a *amendOperationAddIndex) genMutations(ctx context.Context, sctx sessionctx.Context, commitMutations tikv.CommitterMutations,
+	kvMap *rowKvMap, resAddMutations *tikv.CommitterMutations) error {
+	newIdxMutation := &tikv.CommitterMutations{}
+	oldIdxMutation := &tikv.CommitterMutations{}
+	// There should be no duplicate keys in deletedOldIndexKeys and insertedNewIndexKeys.
+	deletedMutations := tikv.NewCommiterMutations(32)
+	insertedMutations := tikv.NewCommiterMutations(32)
 	for i, key := range commitMutations.GetKeys() {
-		keyOp := commitMutations.GetOps()[i]
 		if tablecodec.IsIndexKey(key) || tablecodec.DecodeTableID(key) != a.info.tblInfoAtCommit.Meta().ID {
 			continue
 		}
-		if !isDeleteOp(keyOp) {
-			continue
+		newIdxMutation.Reset()
```

Comment:
We can reuse `deletedMutatiosn` and `insertedMutations` with `Pop` method that remove the last mutation.

Reply:
Or just return a single `Mutation` for `genOldIdxKey` and `genNewIdxKey`.